### PR TITLE
fix: handle invalid url

### PR DESCRIPTION
**Pull request by climba03003**
Fixes: #3127 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by zekth — APPROVED** [review]

File: test/404s.test.js [hunk 1]
```diff
@@ -1738,6 +1738,41 @@ test('400 in case of bad url (pre find-my-way v2.2.0 was a 404)', t => {
...
+      t.equal(response.statusCode, 400)
+      t.same(JSON.parse(response.payload), {
+        error: 'Bad Request',
+        message: "'%c0' is not a valid url component",
+        statusCode: 400
+      })
+    })
+  })
+
+  t.test('Only / is registered', t => {
```
Comment (Eomm) [inline test/404s.test.js:1758]: If we have a route with the parameter what will happen? (such as your example to reproduce the issue)
Is the parameter filled or the 400 is triggered?

**Review by mcollina — APPROVED** [review]
lgtm

File: test/404s.test.js (hunk 1 above)
Comment (climba03003) [reply test/404s.test.js:1758]: 400 is triggered, it is on the test above.

File: test/404s.test.js [hunk 2]
```diff
@@ -1738,6 +1738,41 @@ test('400 in case of bad url (pre find-my-way v2.2.0 was a 404)', t => {
...
+    fastify.inject({
+      url: '/%c0',
+      method: 'GET'
+    }, (err, response) => {
+      t.error(err)
+      t.equal(response.statusCode, 400)
+      t.same(JSON.parse(response.payload), {
+        error: 'Bad Request',
+        message: "'%c0' is not a valid url component",
+        statusCode: 400
```
Comment (jsumners) [inline test/404s.test.js:1753]: Why isn't this a 404? The route for `%c0` hasn't been defined. In fact, `%c0` is completely valid according to https://datatracker.ietf.org/doc/html/rfc1738#section-2.2

Comment (climba03003) [reply test/404s.test.js:1753]: It is possible to throw `404` if we use change `onBadUrl` to `basic404`. Should we change it to `404` according to IETF @mcollina?

Comment (climba03003) [reply test/404s.test.js:1753]: It is possible to throw `404` if we use `basic404` for the `onBadUrl`. Should we change it to throwing 404 according to IETF @mcollina ?

Comment (mcollina) [reply test/404s.test.js:1753]: I would go with a 404.

File: lib/fourOhFour.js [hunk 3]
```diff
@@ -58,6 +58,24 @@ function fourOhFour (options) {
...
+  function onBadUrl (path, req, res) {
+    const { url, method } = req
+    const message = `Route ${method}:${url} not found`
+    const body = `{"error":"Not Found","message":"${message}","statusCode":404}`
+
+    // simulate normal route logging
+    const id = genReqId(req)
+    const childLogger = logger.child({ reqId: id })
+    childLogger.info({ req }, 'incoming request')
+    childLogger.info({ req }, message)
```
Comment (jsumners) [inline lib/fourOhFour.js:70]: What if `disableRequestLogging: true`?

File: lib/fourOhFour.js [hunk 4]
```diff
@@ -58,6 +58,24 @@ function fourOhFour (options) {
     })
   }
 
+  function onBadUrl (path, req, res) {
+    const { url, method } = req
+    const message = `Route ${method}:${url} not found`
+    const body = `{"error":"Not Found","message":"${message}","statusCode":404}`
+
+    // simulate normal route logging
+    const id = genReqId(req)
```
Comment (climba03003) [inline lib/fourOhFour.js:67]: I would like to have some feedback about the logging. Do we have a better approach for this one?

File: lib/fourOhFour.js (hunk 3 above)
Comment (climba03003) [reply lib/fourOhFour.js:70]: Missed this option.

**Review by jsumners — APPROVED** [review]
Looks good to me.

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**